fix(common): ensures hash location strategy includes current path fro…

haydenbr opened the pull request:
…m window.location.pathname

Previously with HashLocationStrategy, any url updates were only made with the hash, for example when the routerLink directives updates the href on anchor tags or in calls to history.pushState or history.replaceState. In the case that a base href is set though, the browser will actually update the url path in the address bar with the path included in the base href. During client side hash routing, this causes unexpected behavior. By always including the current location.pathname in history updates, this ensures that HashRouting will never cause a change to anything before the hash, which is what one would expect.

Fixes #33000

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

Issue Number: #33000

## What is the new behavior?

When using HashRouting, the url before the hash will not be modified

## Does this PR introduce a breaking change?

- [ ] Yes
- [ x] No

Commit:
fix(common): ensures hash location strategy includes current path from window.location.pathname

Previously with HashLocationStrategy, any url updates were only made with the hash, for example when the routerLink directives updates the href on anchor tags or in calls to history.pushState or history.replaceState. In the case that a base href is set though, the browser will actually update the url path in the address bar with the path included in the base href. During client side hash routing, this causes unexpected behavior. By always including the current location.pathname in history updates, this ensures that HashRouting will never cause a change to anything before the hash, which is what one would expect.

Fixes #33000

mhevery commented:
Could we have some tests for this?

haydenbr commented:
Sure! Sorry I didn't add those to start.

My wedding is in 1.5 weeks, but will add tests ASAP after that.

jesusbarcenilla commented on packages/common/src/location/hash_location_strategy.ts:
```diff
@@ -61,22 +61,20 @@ export class HashLocationStrategy extends LocationStrategy {
-    return url.length > 0 ? ('#' + url) : url;
+    const path = this._platformLocation.pathname;
+
+    return url.length > 0 ? (path + '#' + url) : path;
```
I think it'd be good to keep the query string too.
```javascript
        const url = this.joinWithSlash(this.baseHref, internal);
        const hash = url.length > 0 ? ('#' + url) : url;
        return this.platformLocation.pathname + this.platformLocation.search + hash;
```

haydenbr replied:
Thanks for reviewing, but I don't think this is necessary. If you have a query string after the hash (the more likely case) then it's included in window.location.hash as it's after the hash, and these are already preserved.

window.location.search would only be non-empty if for some reason you had a queryString _before_ the hash in your url, which seams extremely unlikely. Currently, these wouldn't have been preserved in the url anyways, and theoretically could be added easily enough in the future if there was a use case for it.

jesusbarcenilla replied:
In my case, I have to deploy my app in Salesforce and I need the query string to control parts of the page that are outside of my angular app. It could be a problem for any other scenario where it is not possible to change the behavior of the server or the angular app is not the only thing in the page.
I think the hash location strategy should not change any part of the url before the hash. I'm using my own strategy at the moment but I really appreciate if you can reconsider to include it. It is a small change and i think it is not going to cause any harm.
Thanks

haydenbr replied:
Ah ok. That makes sense. And if the query string isn't defined, it's just adding an empty string anyways so no harm done. I'll include that as well. I ended up having to include the original host as well.

Would you mind helping me with testing if you get a chance? There's two issues I'm running into with testing:

1. Actually writing tests for the original change
2. After a small change I had to make today, one existing test case is failing: 'WebWorker Router should route on click'

If you were able to help out with either of these, it'd be super cool!

jesusbarcenilla replied:
Sure. I'll have a look.

Commit:
fix(common): update prepareExternalUrl for hash strategy and update fallback naivgation with window.location

For hash location strategy, update prepareExternalUrl to maintain both the existing location.origin and location.pathname. This is requred to ensure that everything before the hash in the url is unchanged. If the baseHref specifies an origin different from the origin in the current url, the browser throws an error due to attempting to navigate to a url with a different origin.

If the History API is not available, then window.location is used. In BrowserPlatformLocation, update fallback navigation for pushState() and replaceState(), to use window.location.assign() and window.location.replace(), respectively. Previously, the fallback location set window.location.href for both, which would only work when routing with relative urls. The assign() and replace() methods work for both relative and absolute urls.

petebacondarwin commented:
@haydenbr - I hope your wedding went well! Are you interested in continuing with this PR?

haydenbr commented:
It went well. Thanks Pete!

I’m not interested at this point. Sorry. I meant to close this a while ago. Ideally, it’s something that should be fixed because the current behavior produces unexpected results. But I got a little overwhelmed with how to add tests and wasn’t sure who to reach out to for help.

If you are willing to help point me in the right direction or put me in touch with someone who can donate a few minutes to help, I should be able to add tests for this to finish it up. If that’s the case, I have a fair free time this week to wrap this up.

________________________________
From: Pete Bacon Darwin <notifications@github.com>
Sent: Tuesday, July 14, 2020 10:54 AM
To: angular/angular <angular@noreply.github.com>
Cc: Hayden Braxton <haydenbraxton@haydenbraxton.com>; Mention <mention@noreply.github.com>
Subject: Re: [angular/angular] fix(common): ensures hash location strategy includes current path fro… (#33007)


@haydenbr<https://github.com/haydenbr> - I hope your wedding went well! Are you interested in continuing with this PR?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/angular/angular/pull/33007#issuecomment-658227522>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFBXCY3AASTA6INZUYJGGIDR3RWRNANCNFSM4I5S3AJA>.

petebacondarwin commented:
Let's do it @haydenbr - it would be great for you to get this in under your name. I am happy to help out. I'll make some suggestions about tests later today.

haydenbr commented:
Great! Thanks. I’ll open the code back up today to see where I was struggling so I have specific questions to ask.

________________________________
From: Pete Bacon Darwin <notifications@github.com>
Sent: Thursday, July 16, 2020 3:49 AM
To: angular/angular <angular@noreply.github.com>
Cc: Hayden Braxton <haydenbraxton@haydenbraxton.com>; Mention <mention@noreply.github.com>
Subject: Re: [angular/angular] fix(common): ensures hash location strategy includes current path fro… (#33007)


Let's do it @haydenbr<https://github.com/haydenbr> - it would be great for you to get this in under your name. I am happy to help out. I'll make some suggestions about tests later today.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/angular/angular/pull/33007#issuecomment-659225068>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFBXCYZ6U3D6HIV4XJBPOM3R32WIHANCNFSM4I5S3AJA>.

petebacondarwin commented:
@haydenbr - before you get stuck in, we should chat with @atscott about this. (I don't want to waste your time).

I was looking into your fix (and the issue that you opened) and I saw that Andrew had closed the issue as a duplicate of two other issues. One was a feature, which he was not keen on, and one is an issue, which I think really is the same as this, which he has marked as a confirmed bug.

So I think we should double check with him, since it will impact the router, which @atscott is dealing with at the moment. I'll ping him when he comes online.

atscott commented:
After further investigation, the current behavior is working as intended and follows [RFC 3986, section 5.2.2: Transform References](https://tools.ietf.org/html/rfc3986#section-5.2.2). There are three possible solutions for preserving query params when using `HashLocationStrategy`:

* Remove the `base href` from the header. Angular will not modify anything before the hash when using `HashLocationStrategy` when if `base href` is not defined. 
* Put the query params in the base href, i.e. `<base href="http://localhost:4200/?param=value">`. When no path or query is defined for a URI (this would be true when using `HashLocationStrategy`), the browser inherits those values from the base
* Provide your own `LocationStrategy` and/or `PlatformLocation` if the default Angular behavior does not cover your use-case well enough.
